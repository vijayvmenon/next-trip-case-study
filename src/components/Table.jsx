/*eslint-disable */
import React, { Fragment, useContext } from 'react';
import PropTypes from 'prop-types';
import { createUseStyles } from 'react-jss';
import constants, { colors } from '../constants';
import { Context } from '../context';
import { ReactComponent as Expand } from '../assets/plus-circle.svg';
import { ReactComponent as Collapse } from '../assets/minus-circle.svg';
import { ReactComponent as CloseIcon } from '../assets/close-icon.svg';
const useStyles = createUseStyles({
  root: {
    tableLayout: 'fixed',
    border: '1px solid black',
    width: '100%',
    borderSpacing: 0,
    boxSizing: 'border-box',
  },
  header: {
    background: colors.BACKGROUND_2,
    color: colors.WHITE,
    padding: 12,
  },
  tableCell: {
    padding: 16,
    borderBottom: '1px solid black',
  },
  icon: {
    widht: 32,
    height: 32,
    color: colors.BACKGROUND_2,
    cursor: 'pointer',
  },
  expandIcon: {
    extend: 'icon',
  },
  collapseIcon: { extend: 'icon' },
  closeIcon: { extend: 'icon', cursor: 'default' },
  iconWithText: {
    paddingLeft: 8,
    display: 'flex',
    alignItems: 'center',
  },
  noDataRow: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
  },
  iconText: {
    paddingLeft: 8,
    fontWeight: 500,
    fontSize: '1rem',
    color: colors.BACKGROUND_2,
  },
  iconTextNoData: {
    paddingLeft: 8,
    fontWeight: 500,
    fontSize: '1.2rem',
    letterSpacing: 2,
    color: colors.TEXT_COLOR,
  },
});
export default function Table({ rows, columns }) {
  const classes = useStyles();
  const { state: { tableExpanded = false } = {}, dispatch } =
    useContext(Context);

  const rowsToShow = rows.length > 5 ? rows.slice(0, 5) : rows;
  const rowsToHide = rows.length > 5 ? rows.slice(5) : [];
  console.log(rowsToShow, rowsToHide, tableExpanded);

  const TableHeader = () => (
    <thead className={classes.header}>
      <tr>
        {columns.map((val) => (
          <th
            key={val.id}
            className={classes.header}
            style={{ textAlign: val?.textAlign }}
          >
            {val.value}
          </th>
        ))}
      </tr>
    </thead>
  );

  const RowsIfCollapsed = () =>
    rowsToShow.map((rowValue, key) => (
      <tr key={key}>
        {columns.map((columnValue, columnKey) => (
          <td
            key={columnKey}
            className={classes.tableCell}
            style={{ textAlign: columnValue.textAlign }}
          >
            {columnValue?.formatter
              ? columnValue.formatter(rowValue?.[columnValue.id], rowValue)
              : rowValue[columnValue?.id]}
          </td>
        ))}
      </tr>
    ));

  const RowsIfExpanded = () =>
    tableExpanded &&
    rowsToHide.length > 0 && (
      <>
        {rowsToHide.map((rowValue, key) => (
          <tr key={key}>
            {columns.map((columnValue, columnKey) => (
              <td
                key={columnKey}
                className={classes.tableCell}
                style={{ textAlign: columnValue.textAlign }}
              >
                {columnValue?.formatter
                  ? columnValue.formatter(rowValue?.[columnValue.id], rowValue)
                  : rowValue[columnValue?.id]}
              </td>
            ))}
          </tr>
        ))}
      </>
    );

  const ExpandRow = () =>
    !tableExpanded &&
    rowsToHide.length > 0 && (
      <tr>
        <td>
          <div className={classes.iconWithText}>
            <Expand
              className={classes.expandIcon}
              onClick={() => dispatch({ type: 'TABLE_EXPAND' })}
            />
            <p className={classes.iconText}>{constants.SHOW_MORE}</p>
          </div>
        </td>
      </tr>
    );

  const CollapseRow = () =>
    tableExpanded &&
    rowsToHide.length > 0 && (
      <tr>
        <td>
          <div className={classes.iconWithText}>
            <Collapse
              className={classes.collapseIcon}
              onClick={() => dispatch({ type: 'TABLE_COLLAPSE' })}
            />
            <p className={classes.iconText}>{constants.SHOW_LESS}</p>
          </div>
        </td>
      </tr>
    );

  const NoDataRow = () =>
    rows.length === 0 && (
      <tr>
        <td colSpan={columns.length}>
          <div className={classes.noDataRow}>
            <CloseIcon className={classes.closeIcon} />
            <p className={classes.iconTextNoData}>{constants.STOP_CLOSED}</p>
          </div>
        </td>
      </tr>
    );

  return (
    <table className={classes.root}>
      <TableHeader />
      <tbody>
        <NoDataRow />
        <RowsIfCollapsed />
        <ExpandRow />
        <RowsIfExpanded />
        <CollapseRow />
      </tbody>
    </table>
  );
}

Table.propTypes = {
  rows: PropTypes.arrayOf(PropTypes.any).isRequired,
  columns: PropTypes.arrayOf(PropTypes.any).isRequired,
};
