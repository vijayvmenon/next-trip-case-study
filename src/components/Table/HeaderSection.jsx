import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import { createUseStyles } from 'react-jss';
import { colors } from '../../constants';
import { Context } from '../../context';

const useStyles = createUseStyles({
  root: {
    display: 'flex',
    flexDirection: 'column',
    padding: '0px 16px',
    '@media(min-width: 768px)': {
      flexDirection: 'row',
      background: colors.TABLE_BACKGROUND,
      color: colors.TEXT_COLOR,
    },
    flexWrap: 'wrap',
    background: colors.BACKGROUND_5,
    justifyContent: 'space-between',
    borderTopLeftRadius: 10,
    borderTopRightRadius: 10,
    color: colors.WHITE,
  },
  stopId: {
    display: 'flex',
    alignItems: 'center',
  },
  headerText: {
    margin: '8px 0px',
    '@media(min-width: 768px)': {
      margin: '1rem 0rem',
    },
  },
  stopIdText: {
    fontWeight: 'bold',
  },
  stopText: {
    margin: '8px 0px',
    '@media(min-width: 768px)': {
      margin: '1rem',
    },
  },
});

export default function TableHeader({ headerText, stopId }) {
  const classes = useStyles();
  const { state: { translations = {} } = {} } = useContext(Context);
  return (
    <div className={classes.root}>
      <h3 className={classes.headerText}>{headerText}</h3>
      <div className={classes.stopId}>
        <div className={classes.stopIdText}>{translations.STOP_HASH}</div>
        <p className={classes.stopText}>{stopId}</p>
      </div>
    </div>
  );
}

TableHeader.defaultProps = {
  stopId: 0,
  headerText: '',
};

TableHeader.propTypes = {
  headerText: PropTypes.string,
  stopId: PropTypes.number,
};
