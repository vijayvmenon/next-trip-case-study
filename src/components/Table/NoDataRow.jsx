import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import { createUseStyles } from 'react-jss';
import { colors } from '../../constants';
import { ReactComponent as CloseIcon } from '../../assets/close-icon.svg';
import { Context } from '../../context';

const useStyles = createUseStyles({
  icon: {
    widht: 32,
    height: 32,
    color: colors.TEXT_COLOR_2,
  },
  closeIcon: { extend: 'icon', color: colors.TEXT_COLOR, cursor: 'default' },
  noDataRow: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
  },
  iconTextNoData: {
    paddingLeft: 8,
    fontWeight: 500,
    fontSize: '1.2rem',
    letterSpacing: 2,
    color: colors.TEXT_COLOR,
  },
});

export default function NoDataRow({ rows, columns }) {
  const classes = useStyles();
  const { state: { translations = {} } = {} } = useContext(Context);

  return (
    rows.length === 0 && (
      <tr>
        <td colSpan={columns.length}>
          <div className={classes.noDataRow} data-testid="no-data-in-table">
            <CloseIcon className={classes.closeIcon} />
            <p className={classes.iconTextNoData}>{translations.STOP_CLOSED}</p>
          </div>
        </td>
      </tr>
    )
  );
}

NoDataRow.propTypes = {
  columns: PropTypes.arrayOf(PropTypes.any).isRequired,
  rows: PropTypes.arrayOf(PropTypes.any).isRequired,
};
