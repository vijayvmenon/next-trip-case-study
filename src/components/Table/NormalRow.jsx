import React from 'react';
import PropTypes from 'prop-types';
import uuid from 'react-uuid';
import { createUseStyles } from 'react-jss';
import { colors } from '../../constants';

const useStyles = createUseStyles({
  tableCell: {
    padding: 16,
    color: colors.TEXT_COLOR,
    fontWeight: 'bold',
    fontSize: '1.1rem',
    // borderBottom: '1px solid black',
  },
  tableCellDescription: {
    padding: 16,
    color: colors.TEXT_COLOR,
    fontSize: '1.1rem',
    // borderBottom: '1px solid black',
  },
});

export default function ExpandedRow({ columns, rowValue }) {
  const classes = useStyles();

  return columns.map((columnValue) => (
    <td
      key={uuid()}
      className={
        columnValue.id === 'description'
          ? classes.tableCellDescription
          : classes.tableCell
      }
      style={{ textAlign: columnValue.textAlign }}
    >
      {columnValue?.formatter
        ? columnValue.formatter(rowValue?.[columnValue.id], rowValue)
        : rowValue[columnValue?.id]}
    </td>
  ));
}

ExpandedRow.propTypes = {
  columns: PropTypes.arrayOf(PropTypes.any).isRequired,
  rowValue: PropTypes.objectOf(PropTypes.any),
};
