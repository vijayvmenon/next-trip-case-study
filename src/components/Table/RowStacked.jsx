import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import uuid from 'react-uuid';
import { createUseStyles } from 'react-jss';
import { colors } from '../../constants';
import { Context } from '../../context';

const useStyles = createUseStyles({
  stackedRoot: {
    display: 'flex',
    alignItems: 'center',
    padding: 8,
    justifyContent: 'space-between',
    marginTop: -10,
    marginBottom: -10,
    '@media(max-width: 500px)': {
      flexWrap: 'wrap',
      padding: '12px 8px',
    },
  },
  label: {
    color: colors.TEXT_COLOR,
    lineHeight: '1.5rem',
    letterSpacing: 3,
    marginRight: '1rem',
    flexShrink: 0,
    '@media(max-width: 500px)': {
      marginRight: '100rem', // this is to wrap all the labels and text at width < 500px
    },
    width: '35%',
  },
  tableCellText: {
    color: colors.TEXT_COLOR,
    fontWeight: 'bold',
    fontSize: '1.1rem',
    margin: 0,
    flexGrow: 1,
  },
  tableTd: {
    padding: '0.875rem',
    width: '100%',
  },
});

export default function RowStacked({ columns, rowValue }) {
  const classes = useStyles();
  const { state: { translations = {} } = {} } = useContext(Context);
  return (
    <td className={classes.tableTd}>
      {columns.map((columnValue) => (
        <div className={classes.stackedRoot} key={uuid()}>
          <label className={classes.label} htmlFor={columnValue?.id}>
            {translations[columnValue.value]}
          </label>
          <p className={classes.tableCellText} id={columnValue?.id}>
            {columnValue?.formatter
              ? columnValue.formatter(rowValue?.[columnValue.id], rowValue)
              : rowValue[columnValue?.id]}
          </p>
        </div>
      ))}
    </td>
  );
}

RowStacked.propTypes = {
  columns: PropTypes.arrayOf(PropTypes.any).isRequired,
  rowValue: PropTypes.objectOf(PropTypes.any).isRequired,
};
