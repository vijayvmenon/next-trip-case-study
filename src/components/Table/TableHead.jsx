import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import { createUseStyles } from 'react-jss';
import { colors } from '../../constants';
import { Context } from '../../context';

const useStyles = createUseStyles({
  header: {
    background: colors.BACKGROUND_5,
    color: colors.WHITE,
    padding: 12,
    lineHeight: '1.5rem',
    letterSpacing: 3,
  },
});

export default function TableHeader({ columns, normalTable }) {
  const classes = useStyles();
  const { state: { translations = {} } = {} } = useContext(Context);
  return normalTable ? (
    <thead className={classes.header}>
      <tr>
        {columns.map((val) => (
          <th
            key={val.id}
            className={classes.header}
            style={{ textAlign: val?.textAlign }}
          >
            {translations[val.value]}
          </th>
        ))}
      </tr>
    </thead>
  ) : null;
}

TableHeader.propTypes = {
  columns: PropTypes.arrayOf(PropTypes.any).isRequired,
  normalTable: PropTypes.bool.isRequired,
};
