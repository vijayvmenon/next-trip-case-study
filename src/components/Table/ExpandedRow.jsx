import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import clsx from 'clsx';
import uuid from 'react-uuid';
import { createUseStyles } from 'react-jss';
import { colors } from '../../constants';
import { ReactComponent as Collapse } from '../../assets/minus-circle.svg';
import NormalRow from './NormalRow';
import RowStacked from './RowStacked';
import { Context } from '../../context';

const useStyles = createUseStyles({
  tableRow: {
    borderTop: '1px solid black',
  },
  icon: {
    widht: 32,
    height: 32,
    color: colors.TEXT_COLOR_2,
  },
  collapseIcon: {
    extend: 'icon',
  },
  clickable: {
    cursor: 'pointer',
  },
  iconWithText: {
    paddingLeft: 8,
    display: 'flex',
    alignItems: 'center',
  },
  iconText: {
    paddingLeft: 8,
    fontWeight: 500,
    fontSize: '1rem',
    color: colors.TEXT_COLOR,
  },
});

export default function ExpandedRow({
  tableExpanded,
  rowsToHide,
  columns,
  click,
  normalTable,
}) {
  const classes = useStyles();
  const { state: { translations = {} } = {} } = useContext(Context);

  return (
    <>
      {tableExpanded && rowsToHide.length > 0 && (
        <>
          {rowsToHide.map((rowValue) => (
            <tr key={uuid()} className={classes.tableRow}>
              {normalTable ? (
                <NormalRow columns={columns} rowValue={rowValue} />
              ) : (
                <RowStacked columns={columns} rowValue={rowValue} />
              )}
            </tr>
          ))}
        </>
      )}
      {tableExpanded && rowsToHide.length > 0 && (
        <tr className={clsx(classes.tableRow, classes.clickable)}>
          <td colSpan={columns.length}>
            {/* eslint-disable-next-line jsx-a11y/click-events-have-key-events */}
            <div
              role="button"
              tabIndex="0"
              className={classes.iconWithText}
              onClick={click}
            >
              <Collapse className={classes.collapseIcon} />
              <p className={classes.iconText}>{translations.SHOW_LESS}</p>
            </div>
          </td>
        </tr>
      )}
    </>
  );
}

ExpandedRow.propTypes = {
  columns: PropTypes.arrayOf(PropTypes.any).isRequired,
  tableExpanded: PropTypes.bool.isRequired,
  rowsToHide: PropTypes.arrayOf(PropTypes.any).isRequired,
  click: PropTypes.func.isRequired,
  normalTable: PropTypes.bool.isRequired,
};
