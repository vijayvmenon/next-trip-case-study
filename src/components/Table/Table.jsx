import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import { createUseStyles } from 'react-jss';
import { useMediaPredicate } from 'react-media-hook';
import { colors } from '../../constants';
import { Context } from '../../context';
import TableHead from './TableHead';
import ExpandedRow from './ExpandedRow';
import CollapsedRow from './CollapsedRow';
import NoDataRow from './NoDataRow';

const useStyles = createUseStyles({
  tableContainer: {
    '@media(min-width:1200px)': {
      width: 1140,
    },
    '@media(max-width:1200px)': {
      width: 800,
    },
    '@media(max-width:900px)': {
      width: 'auto',
    },
    '@media(max-width:768px)': {
      width: 'auto',
    },
  },
  root: {
    tableLayout: 'fixed',
    width: '100%',
    borderSpacing: 0,
    boxSizing: 'border-box',
    borderCollapse: 'collapse',
    background: colors.TABLE_BACKGROUND,
  },
});

export default function Table({ rows, columns }) {
  const classes = useStyles();
  // this will show normal table at screen width > 768 px and responsive table at screen width < 768px
  const normalTable = useMediaPredicate('(min-width: 768px)');
  const { state: { tableExpanded = false } = {}, dispatch } =
    useContext(Context);

  const rowsToShow = rows.length > 5 ? rows.slice(0, 5) : rows;
  const rowsToHide = rows.length > 5 ? rows.slice(5) : [];

  return (
    <div className={classes.tableContainer}>
      <table className={classes.root}>
        <TableHead columns={columns} normalTable={normalTable} />
        <tbody>
          <NoDataRow rows={rows} columns={columns} />
          <CollapsedRow
            rowsToShow={rowsToShow}
            columns={columns}
            click={() => dispatch({ type: 'TABLE_EXPAND' })}
            rowsToHide={rowsToHide}
            tableExpanded={tableExpanded}
            normalTable={normalTable}
          />
          <ExpandedRow
            tableExpanded={tableExpanded}
            rowsToHide={rowsToHide}
            columns={columns}
            click={() => dispatch({ type: 'TABLE_COLLAPSE' })}
            normalTable={normalTable}
          />
        </tbody>
      </table>
    </div>
  );
}

Table.propTypes = {
  rows: PropTypes.arrayOf(PropTypes.any).isRequired,
  columns: PropTypes.arrayOf(PropTypes.any).isRequired,
};
