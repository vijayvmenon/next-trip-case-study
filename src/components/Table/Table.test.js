import { render, fireEvent, screen } from '@testing-library/react';
import Table from './Table';
import renderer from 'react-test-renderer';
import { Context } from '../../context';
import Rows from '../../mockData/rows.json';

const dispatch = jest.fn(); // mock dispatch

const columns = [
  { id: 'route_short_name', value: 'ROUTE', textAlign: 'left', key: '1' },
  { id: 'description', value: 'DESTINATION', textAlign: 'left', key: '2' },
  {
    id: 'departure_text',
    value: 'DEPARTS',
    textAlign: 'right',
    key: '3',
  },
];

const lessRows = Rows.departures;
const noRows = [];
const moreRows = Rows.departures2;

describe('Component:Table.jsx', () => {
  test('Table.jsx renders correctly', () => {
    const table = renderer
      .create(
        <Context.Provider value={{ dispatch }}>
          <Table columns={columns} rows={lessRows} />,
        </Context.Provider>,
      )
      .toJSON();
    expect(table).toMatchSnapshot();
  });

  test('table displays passed rows', () => {
    render(
      <Context.Provider value={{ dispatch }}>
        <Table rows={lessRows} columns={columns} />,
      </Context.Provider>,
    );
  });

  test('table is not displayed if rows are empty', () => {
    const { getByTestId } = render(
      <Context.Provider value={{ dispatch }}>
        <Table rows={noRows} columns={columns} />,
      </Context.Provider>,
    );
    expect(getByTestId('no-data-in-table')).toBeTruthy();
  });

  test('Expand Icon is showed when there are more than 5 rows', () => {
    const { getByTestId } = render(
      <Context.Provider value={{ dispatch }}>
        <Table rows={moreRows} columns={columns} />,
      </Context.Provider>,
    );
    expect(getByTestId('table-expand-icon')).toBeTruthy();
  });
});
