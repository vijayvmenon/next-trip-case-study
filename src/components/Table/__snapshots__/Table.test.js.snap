// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component:Table.jsx Table.jsx renders correctly 1`] = `
Array [
  <div
    className="tableContainer-0-2-1"
  >
    <table
      className="root-0-2-2"
    >
      <tbody>
        <tr
          className="tableRow-0-2-8"
        >
          <td
            className="tableTd-0-2-17"
          >
            <div
              className="stackedRoot-0-2-14"
            >
              <label
                className="label-0-2-15"
                htmlFor="route_short_name"
              />
              <p
                className="tableCellText-0-2-16"
                id="route_short_name"
              >
                Blue
              </p>
            </div>
            <div
              className="stackedRoot-0-2-14"
            >
              <label
                className="label-0-2-15"
                htmlFor="description"
              />
              <p
                className="tableCellText-0-2-16"
                id="description"
              >
                to Mpls-Target Field
              </p>
            </div>
            <div
              className="stackedRoot-0-2-14"
            >
              <label
                className="label-0-2-15"
                htmlFor="departure_text"
              />
              <p
                className="tableCellText-0-2-16"
                id="departure_text"
              >
                Due
              </p>
            </div>
          </td>
        </tr>
        <tr
          className="tableRow-0-2-8"
        >
          <td
            className="tableTd-0-2-17"
          >
            <div
              className="stackedRoot-0-2-14"
            >
              <label
                className="label-0-2-15"
                htmlFor="route_short_name"
              />
              <p
                className="tableCellText-0-2-16"
                id="route_short_name"
              >
                Blue
              </p>
            </div>
            <div
              className="stackedRoot-0-2-14"
            >
              <label
                className="label-0-2-15"
                htmlFor="description"
              />
              <p
                className="tableCellText-0-2-16"
                id="description"
              >
                to Mall of America
              </p>
            </div>
            <div
              className="stackedRoot-0-2-14"
            >
              <label
                className="label-0-2-15"
                htmlFor="departure_text"
              />
              <p
                className="tableCellText-0-2-16"
                id="departure_text"
              >
                7 Min
              </p>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
  ",",
]
`;
